### PR TITLE
drivers: hci: da1469x: Add driver for CMAC core on DA1469x

### PR DESCRIPTION
**Pull request by andrzej-kaczmarek**
This adds HCI driver which enables communication with CMAC core on
Renesas SmartBond(tm) DA1469x series. The CMAC firmware is provided as
a library via hal_renesas along with APIs to communicate via shared
memory. The protocol over shared memory mailboxes is H4 so the code is
based on existing hci_h4 driver.

### CODE REVIEW

**Comment by nordicjm** [comment]
> The controller image is provided in hal_renesas repository

So that's a blob? This needs to follow https://docs.zephyrproject.org/latest/contribute/bin_blobs.html and have TSC approval

**Comment by carlescufi** [comment]
> > So that's a blob? This needs to follow [docs.zephyrproject.org/latest/contribute/bin_blobs.html](https://docs.zephyrproject.org/latest/contribute/bin_blobs.html) and have TSC approval
> 
> _Binary blobs (or blobs for short) are files containing proprietary machine code or data in a binary format, e.g. without corresponding source code released under an OSI approved license._
> 
> So if I understand the above correctly it does not apply here since included image is a prebuilt Apache NimBLE controller. The source code is publicly available and licensed under Apache 2.0 license. It's exactly the same image that newt tool builds and links when building Apache Mynewt for DA1469x.

This is an edge case then. You are still providing a precompiled library in binary form, so we need to at least discuss it in the TSC.

**Comment by nordicjm** [comment]
Note that another vendor has done similar previously and it was submitted as a blob: https://github.com/zephyrproject-rtos/zephyr/issues/54534

**Comment by andrzej-kaczmarek** [comment]
> > _Binary blobs (or blobs for short) are files containing proprietary machine code or data in a binary format, e.g. without corresponding source code released under an OSI approved license._
> > So if I understand the above correctly it does not apply here since included image is a prebuilt Apache NimBLE controller. The source code is publicly available and licensed under Apache 2.0 license. It's exactly the same image that newt tool builds and links when building Apache Mynewt for DA1469x.
> 
> This is an edge case then. You are still providing a precompiled library in binary form, so we need to at least discuss it in the TSC.

Understood. Do you want me to create an issue for binary blob anyway or you will discuss it first as it is now? I can also add some kind of readme to PR in hal_renesas with steps to rebuild that binary.

**Comment by carlescufi** [comment]
> Note that another vendor has done similar previously and it was submitted as a blob: #54534

But I am not sure that the source code for those Infineon libraries was available under an open source license actually, was it? @ifyall @npal-cy

**Comment by carlescufi** [comment]
> Understood. Do you want me to create an issue for binary blob anyway or you will discuss it first as it is now? I can also add some kind of readme to PR in hal_renesas with steps to rebuild that binary.

Let's create the issue for a binary blob @andrzej-kaczmarek, because precompiling an open source project is new.  If you do this today we can include it in today's TSC.

> The reason why we want to provide that prebuilt binary is to make BLE work on DA1469x ootb, otherwise people will need to get complete Mynewt env just to create binary.

I know this is not what you propose, but what about using the Zephyr LL instead?

**Comment by andrzej-kaczmarek** [comment]
> > Understood. Do you want me to create an issue for binary blob anyway or you will discuss it first as it is now? I can also add some kind of readme to PR in hal_renesas with steps to rebuild that binary.
> 
> Let's create the issue for a binary blob @andrzej-kaczmarek, because precompiling an open source project is new. If you do this today we can include it in today's TSC.

Done: https://github.com/zephyrproject-rtos/zephyr/issues/57987

> > The reason why we want to provide that prebuilt binary is to make BLE work on DA1469x ootb, otherwise people will need to get complete Mynewt env just to create binary.
> 
> I know this is not what you propose, but what about using the Zephyr LL instead?

Porting Zephyr LL to CMAC architecture would require lots of work so for now I'm afraid that's not an option.

**Comment by carlescufi** [comment]
@andrzej-kaczmarek @sjanc the HAL PR is now merged, please update with SHA.

**Comment by andrzej-kaczmarek** [comment]
@carlescufi we'll need to make some changes after power management code is merged (https://github.com/zephyrproject-rtos/zephyr/pull/61857) so I'll update with final version once we have that done

**Comment by jhedberg** [comment]
@andrzej-kaczmarek please fix the compliance check failures

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:65]: Unless you've profiled the code and concluded that the compiler doesn't do an optimal job, please don't do these explicit inline declarations. Same goes for the other functions here (I won't comment on them separately).

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:451]: To be able to safely do this, the driver first needs to make sure it requests sufficient headroom in the buffers:
https://github.com/zephyrproject-rtos/zephyr/blob/eddb3822d2aa746026837252b3e337b0c4ef067a/subsys/bluetooth/host/Kconfig#L75-L93

I know this is not the prettiest solution, since that list of defaults keeps growing. Once the HCI driver interface has been converted to a proper Zephyr driver API (see #51503), this information could potentially come from a devicetree property.

That said, how much have you actually tested this driver? Without requesting the buffer headroom you should be getting a buffer underflow and potential memory corruption.

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:388]: Maybe add code comments explaining where the calls to these functions are coming from, since they're not in the main Zephyr tree.

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:475]: Not directly related to this PR, but it seems a bit arbitrary that `close()` would depend on the Zephyr host stack. The only other user of the HCI driver API right now. is HCI_RAW, which indeed doesn't use `close()`, but it seems wrong to me that the HCI driver API makes such assumptions about the higher level implementation that uses the API. I guess we can fix this once we move to a native Zephyr driver API.

**Review by jhedberg — CHANGES_REQUESTED** [review]

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [reply drivers/bluetooth/hci/hci_da1469x.c:453]: > Seems like BT_HCI_RESERVE was removed in 3.6 and headroom is now always reserved.

Yes, you can disregard this comment thanks to the removal.

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:33]: These are all available in `include/zephyr/bluetooth/hci_types.h` please use those defines instead.

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:103]: "UART" in the log message seems wrong, since you're not calling a UART API

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:254]: same here

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [inline drivers/bluetooth/hci/hci_da1469x.c:298]: and here

**Review by jhedberg — CHANGES_REQUESTED** [review]
A few more minor comments. Also, please pay attention to the work done in #72323. This driver will also need to be converted as part of that other PR (assuming this PR gets merged first).

**Comment by andrzej-kaczmarek** [comment]
west.yml is now updated with proper hash for HAL

File: drivers/bluetooth/hci/hci_da1469x.c
Comment (jhedberg) [reply drivers/bluetooth/hci/hci_da1469x.c:471]: @andrzej-kaczmarek could I get some comment here? Drivers shouldn't generally know or care who their user is (the normal bluetooth host stack or HCI_RAW). So I think it might be cleaner to just remove the #ifdefs from close().

**Comment by jhedberg** [comment]
@andrzej-kaczmarek the PR looks generally fine to me. I'd be willing to approve if it wasn't for the open question regarding the #ifdefs around close().

**Review by blauret — APPROVED** [review]

**Comment by blauret** [comment]
@jhedberg, I think your last comment was addressed.

**Review by jhedberg — APPROVED** [review]

**Pull request merged**
